server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,255 @@
+<?php
+/**
+ * Server-Timing API: Perflab_Server_Timing class
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Class controlling the Server-Timing header.
+ *
+ * @since n.e.x.t
+ */
+class Perflab_Server_Timing {
+
+	/**
+	 * Map of registered metric slugs and their metric instances.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private $registered_metrics = array();
+
+	/**
+	 * Map of registered metric slugs and their registered data.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private $registered_metrics_data = array();
+
+	/**
+	 * Registers a metric to calculate for the Server-Timing header.
+	 *
+	 * This method must be called before the {@see 'perflab_server_timing_send_header'} hook.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $metric_slug The metric slug.
+	 * @param array  $args        {
+	 *     Arguments for the metric.
+	 *
+	 *     @type callable $measure_callback The callback that initiates calculating the metric value. It will receive
+	 *                                      the Perflab_Server_Timing_Metric instance as a parameter, in order to set
+	 *                                      the value when it has been calculated. Metric values must be provided in
+	 *                                      milliseconds.
+	 *     @type string   $access_cap       Capability required to view the metric. If this is a public metric, this
+	 *                                      needs to be set to "exist".
+	 * }
+	 */
+	public function register_metric( $metric_slug, array $args ) {
+		if ( isset( $this->registered_metrics[ $metric_slug ] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: metric slug */
+				sprintf( __( 'A metric with the slug %s is already registered.', 'performance-lab' ), $metric_slug ),
+				''
+			);
+			return;
+		}
+
+		if ( did_action( 'perflab_server_timing_send_header' ) && ! doing_action( 'perflab_server_timing_send_header' ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: WordPress action name */
+				sprintf( __( 'The method must be called before or during the %s action.', 'performance-lab' ), 'perflab_server_timing_send_header' ),
+				''
+			);
+			return;
+		}
+
+		$args = wp_parse_args(
+			$args,
+			array(
+				'measure_callback' => null,
+				'access_cap'       => null,
```

Comment:
@ankitrox I wanted to intentionally avoid a default here, to make sure developers think about it. Having a default is not always a good idea. If we use a strict default (like `manage_options`), it will lead to most metrics not being visible in the frontend, making it almost useless for field analysis. If we use a loose default (like `exist`), it will lead to sensitive metrics being visible in the frontend when they shouldn't.